File: 16. 3Sum Closest/Dynamic/3sum_closest.js
function sumOfTheResult (list){
	let out = 0;
	
	for (i of list){
		for (j of i){
			out += j;
		}
	}
	
	return out;
}

function absoluteValue(number){
	return (number > 0) ? (number * 1) : (number * -1);
}

function sumOfThreeClosest(list, target = 0){
	let len_list = list.length;
	let out = [];
	let closest = null;
	
	for (let a = 0; a < len_list; a++){
		for (let b = 0; b < len_list; b++){
			if (b > a){
				for (let c = 0; c < len_list; c++){
					if (c > b){
						if ((a != b) && (a != c) && (b !=c)){
							let aa = list[a];
							let bb = list[b];
							let cc = list[c];
							
							////if (($aa != $bb) && ($aa != $cc) && ($bb != $cc)) //toggle this for no similar value elements
							
							if (true){
								let difference = absoluteValue((aa + bb + cc) - target);
								//console.log(String(aa) + "\t" + String(bb) + "\t" + String(cc) + "\t\t" + String(difference) + "\n");
								
								if (closest === null){
									closest = difference;
								}
								
								if (difference == closest){
									//let z = [a, b, c];
									let z = [list[a], list[b], list[c]];
									out.push(z);
								} else if (difference < closest){
									let z = [list[a], list[b], list[c]];
									//let z = [a, b, c];
									out.push(z);
									closest = distance;
								}
							}
						}
					}
				}
			}
		}
	}
	
	return (out);
}

let nums = [-1,2,1,-4];

let closest = sumOfThreeClosest(nums,1);

console.log(JSON.stringify(closest));
console.log(sumOfTheResult(closest));
console.log("\n\n\n");

File: 16. 3Sum Closest/Dynamic/3sum_closest.rb
def absolute_value(number)
	return (number < 0) ? (number * -1) : (number * 1)
end

def sum_of_the_result(list)
	out = 0
	
	list.each do |i|
		i.each do |j|
			out += j
		end
	end
	
	return out
end

def sum_of_three_closest (list, target = 0)
	len_list = list.length
	out = []
	closest = nil
	
	for a in 0...len_list
		for b in 0...len_list
			if (b > a)
				for c in 0...len_list
					if (c > b)
						if (a != b) and (a != c) and (b != c)
							aa = list[a]
							bb = list[b]
							cc = list[c]
							##if (aa != bb) and (aa != cc) and (bb != cc) #Toggle this for no similarr value elements
							if true
								difference = absolute_value((aa + bb + cc) - target)
								if closest == nil
									closest = difference
								end
								
								if (difference == closest)
									#z = [a, b, c]
									z = [list[a], list[b], list[c]]
									out << z
								elsif (difference < closest)
									out = []
									#z = [a, b, c]
									z = [list[a], list[b], list[c]]
									out << z
									closest = difference
								end
							end
						end
					end
				end
			end
		end
	end
	
	return out
end

nums = [-1,2,1,-4]

closest = sum_of_three_closest(nums, 1)

puts(String(closest))
puts(sum_of_the_result(closest))
